### DISCUSSION TITLE
Ideas for improving popup menus across the engine

### DISCUSSION THREAD

**Discussion by mieldepoche** [godotengine/godot-proposals#7629] [Editor]
disclaimer: idk what I'm doing

A recurent problem with the Resource fields in the inspector is how hard it can be to access common operations such as [make unique](https://github.com/godotengine/godot-proposals/issues/7626) and [load/quick load](https://github.com/godotengine/godot-proposals/issues/7454).

the popup contains:
 - new resource options
 - load options
 - resource operations (when the field is not empty)

A common proposal is to reorder them: put common options on top and the variable list of new resource options on the bottom.
However this would harm the experience in the number of cases where you often want to create a new resource in place (collision shape, curve, gradients, etc.).

To sum it up: the popup is bloated. And it's not the only one:
|resource popup|rmb on node popup|
|-|-|
| ![image](https://github.com/godotengine/godot-proposals/assets/71937306/d1f194d7-a76b-4e11-92f6-171563b5cabc) | ![image](https://github.com/godotengine/godot-proposals/assets/71937306/33e5715f-e6d4-408e-9ae4-a579ef197367) |

Here are ideas to rethink the popups in the engine:

# idea 1: more sophisticated popup layouts
something common in other software is popups that are not just a vertical list of string options.

<details>
<summary>Videos of other software</summary>

edit: for some reason those stopped playing on my end :thinking: if it's the case, you can right-click them and open them in a new tab...

|in bitwig studio | in lightworks (notice how the popup can be moved!) | in unreal |
|-|-|-|
| <video src="https://github.com/godotengine/godot-proposals/assets/71937306/16a6fe22-0462-4bfa-9070-2c2668238e83">  | <video src="https://github.com/godotengine/godot-proposals/assets/71937306/0afbeaca-128a-4f1c-8520-e85953f393e5">  | <video src="https://github.com/godotengine/godot-proposals/assets/71937306/28305d71-2085-43a5-8d6f-41b80399af91"> |

</details>

in godot, one could imagine having non-trivial layouts for example, taking inspiration from the above:

<details>
<summary>Godot mockups</summary>

| current | new (using a "broken link" icon for make unique instead of a "duplicate" one that looks a bit too much like a "copy" icon, this was a test) | another mockup, spawned  from an empty `stream` field (notice how the load and quick load options don't really harm the experience if you're trying to create a new resource in place) |
|-|-|-|
| ![mockup01_ref](https://github.com/godotengine/godot-proposals/assets/71937306/8495afab-2fb7-474e-ae42-4f22064b425b) | ![mockup01-2](https://github.com/godotengine/godot-proposals/assets/71937306/16b2837e-fdc1-4ccd-8632-7a265cbca038) | ![mockup02](https://github.com/godotengine/godot-proposals/assets/71937306/7256025a-137e-49b0-b980-07b7cffe8b4b) |

</details>

The benefits of this design idea are that related options are "more" grouped together than if they were so vertically, it saves vertical space and therefore time to find what you're looking for since categories are more clearly separated (maybe debatable but I think so).
The two main menus that would benefit from this is the resource popup in the inspector and the node popup in the scene tree. maybe also the file popup in the FileSystem.

# idea 2 (specific to resource field in inspector): add a `+` button
this is what we're talking about right now:
![image](https://github.com/godotengine/godot-proposals/assets/71937306/1b8aec3d-3f71-415e-aade-e546f21945db)

These buttons all spawn a popup containing things to: create a new resource, copy and paste it, load a new one, clear it.
problem: putting all of this in the same popup is hard, and heavy even with the improvements imaginated in idea 1.

Solution: split the choice *before* the popup:

|an empty field (text in red for information purposes, mockup only) | a field with a resource | a `SphereShape3D` resource, with no preview|
|-|-|-|
| ![image](https://github.com/godotengine/godot-proposals/assets/71937306/b92f09e3-3ae2-4c5f-84c0-5027148bade6) | ![image](https://github.com/godotengine/godot-proposals/assets/71937306/f3d7be1c-9645-452b-b178-eefb94eee64d)  |![image](https://github.com/godotengine/godot-proposals/assets/71937306/8830efc7-8e2e-447e-8196-5cb8198afb53) |

behavior:
- \<empty\> field
  - `+` button opens the list of resources to create a new one
  - `load` button opens the load dialog
  - `quick load` button opens the quick load dialog
  - rmb doesn't do anything?
- filled field
  - `+` button opens the list of resources to create a new one (it's smaller, but you'll need it less often. Another quick way to put a new resource in there is clearing the resource and using the big + button from the \<empty\> state)
  - clicking on the resource enter edit mode, just like it is now
  - the `v` button opens the load/quick load/clear/copy/paste/show in filesystem/etc. menu, stripped of the resource list!
  - rmb opens the same menu as above

**Comment by Mickeon** [comment]
> ![image](https://github.com/godotengine/godot-proposals/assets/66727710/032aeec7-a9b6-46a3-b6bf-b7e909758d78)

A very apparent problem to me with this solution is assuming that only one **Resource** would be visible. With even more than one, I do feel like it would be overwhelming. It's a lot of buttons all equally as likely to be clicked on.
Furthermore, just because those properties are empty (**null**) does not mean they _need_ to be filled with a **Resource**. With these icons, however, it's... as if the icons urge the user to.

I suppose that having the buttons appear only when hovering over them _could_ be a solution. How nicely would all of this work when moving with a keyboard, though?

> ![image](https://github.com/godotengine/godot-proposals/assets/66727710/603969d4-b5d1-401f-b75d-c32e170aaf51)

This is a pretty generous mock-up, too. It actually would look nice. However, the dock can be oftentimes shrunk and tucked to the side. Even when decently expanded, it does currently struggle to display some noteworthy information like the names of nested **Resources**. An additional button _may_ make that issue worse.

**Reply by mieldepoche** [reply]
> I suppose that having the buttons appear only when hovering over them could be a solution. How nicely would all of this work when moving with a keyboard, though?

that would be nice. I'd say tabbing focus through the field would show the buttons as focus enter, and hide them when it exits (focus going from + to load to quick load).

> An additional button may make that issue worse

as you mentionned earlier, only showing it when hovering (and I'll add, only showing the `v` button when hovering as well) would help.

**Reply by mieldepoche** [reply]
quick and dirty mockup

[simplescreenrecorder-2023-09-10_05.06.59.webm](https://github.com/godotengine/godot-proposals/assets/71937306/1de71ce1-0bc8-43c0-9082-42de83e33128)

**Comment by mechPenSketch** [comment]
For reference, here is how the Resource field works (WIP):

**Accessing the Field**
Firstly, to get the field, I call the following from the `EditorPlugin` class: `get_editor_interface().get_inspector().get_children()`. Unfortunately, this is the further I could get, which are an array of `EditorInspectorCategory` and `EditorInspectorSection`. I think the latter needs to be investigated next.

**How ~EditorInspectorSection~ EditorProperty works**
`EditorInspectorSection` is not available in the docs, so I have to look into Godot's code. It can be found in [editor_inspector.cpp](https://github.com/godotengine/godot/blob/master/editor/editor_inspector.cpp). It looks like `EditorProperty` is the one holding the input for a property, while `EditorInspectorSection` contains a group of `EditorProperty`.

**NVM ignore this post**
@aXu-AP has made an edit on [editor_resource_picker.cpp](https://github.com/godotengine/godot/blob/master/editor/editor_resource_picker.cpp) instead.

**Comment by aXu-AP** [comment]
I think this idea is good, altough I would make some changes:
* There's a lack of paste button in your mockup. However, adding fourth button would make things cramped. Moreover, having 2 folder icons might make the user think longer which one to use. I suggest having only ➕ New  and 📂 Quick Load... options on the panel, Load... and Paste in a dropdown menu. Maybe have a redundant quick load also in there, for clarity and for menu to have some content if paste isn't available.
* I would align icons on the right to be less intrusive (and, as Mickeon suggested, show only on mouse hover)
* Keyboard navigation would have 3 targets on that row, but I don't think that's a big problem. There's usually not too many resources on a node for it to make navigation too slow, and even now there's 2 controls per resource (for no reason, both of them open the same menu).

Mockup:
![kuva](https://github.com/godotengine/godot-proposals/assets/1621768/fedc314d-68e5-4c42-9865-c94eb75f562c)

When the resource isn't empty, move the quick load into menu, but leave new icon. Again, right-align the icons.
![kuva](https://github.com/godotengine/godot-proposals/assets/1621768/257be36e-0cde-4453-b836-54d117df6d6e)

**Comment by aXu-AP** [comment]
I tried implementing the resource popup idea so that we can test it in practise: https://github.com/godotengine/godot/pull/91540

**Comment by mechPenSketch** [comment]
Related, but only to field input: I have proposed a new annotation where the drop menu would list the resource files in a directory given as an arguement: https://github.com/godotengine/godot-proposals/issues/9937

**Reply by mechPenSketch** [reply]
I'm now working on a PR: https://github.com/godotengine/godot/pull/103250

**Comment by pixel-boy** [comment]
Maybe add a way to customize the editor, with @export_resource or something like that?

I use custom resources a lot, and being able to access the quick_load menu in 1 click could be handy in some cases, and I wouldn't mind having to create the resources via the resource creation interface.

Resources that would act a bit like @export enums, but whose list would be based on files like quick_load could be really good for workflow in some contexts.

**Comment by mechPenSketch** [comment]
An EditorResourcePicker is added to Resource Field: https://github.com/godotengine/godot/pull/97860

The merge pull is approved for Godot 4.4.

**Reply by arkology** [reply]
Actually it was reverted.

**Reply by mechPenSketch** [reply]
I've looked through PR again and saw there's a [subsequent PR](https://github.com/godotengine/godot/pull/102196) that reverts the original PR I've linked.

**Reply by arkology** [reply]
If anyone is still interested, I have tried to restore the PR functionality, addressing the raised concerns. https://github.com/godotengine/godot/pull/104490
